Replace Worker with Runner

joschahenningsen opened the pull request:
### Motivation and Context

The current worker implementation is hard to maintain, we therefore want to replace it with the new runner. The goal is to have a clean code base that replaces everything the worker does.

### todos:

- All API endpoints the current worker offers should be available in the runner
  - [x] Request Stream
  - [x] Request Stream End 
  - [ ] Generate live preview #1435
  - [ ] Generate Section images #1435
  - [ ] Delete Section images (low prio) #1435
  - [ ] Combine Thumbnails #1435
- All API endpoints TUM-Live currently offers for the runner
  - [x] Register
  - [x] Heartbeat (nice to have)
  - [ ] request selfstream #1437
  - [ ] notify vodUpload finished #1436
  - [ ] Notify Silence results #1436
  - [x] Notify Stream Started #1436
  - [ ] Notify Thumbnails finished #1436
  - [ ] Get stream info for upload #1437
  - [ ] notify Transcoding Failure #1436

Functionality wise, the runner should:
- When request stream is called: 
  - grab the url with ffmpeg and create a hls stream on disk and notify tum-live that the stream is available
  - After the stream is done, convert the files, append them together and upload them to the vod service. 
  - create a thumbnail and the thumnails for the seekbar
  - detect silences in the video and send them to tum-live (not high prio for now)
- when a vod is uploaded, everything that happens after a livestream should also happen to the uploaded file
- when a selfstreamer starts streaming
  - ask tum-live for the details of the token and continue just like with a normal stream

joschahenningsen committed:
start new runner implementation

DawinYurtseven committed:
First Push of the Runner:
- Actions refined and now with postman, working selfstream requests
- A site page development has begun with the needed functionality of heartbeats still in development for the response
- the model/runner.go and dao/runner.go have been edited to work better with existing functionalities

TODO:
- Thumbnail.go still needs tweaking
- rest implementations need to be done
- having the runners be displayed onto the webpage

please text me over element or github if there is anything I forgot to mention in this push

DawinYurtseven committed:
a slight push of progress before switchi

DawinYurtseven committed:
added some communication with tumlive and runner and working on selfstream.go

DawinYurtseven committed:
cmd.yaml update to original

DawinYurtseven committed:
merge runner with dev

DawinYurtseven committed:
Changed transcode command for quality

SebiWrn committed:
Updated runner dependencies

SebiWrn committed:
Fixed cycle import

DawinYurtseven committed:
Import cycle broken for server communications

DawinYurtseven committed:
Merge remote-tracking branch 'origin/enh/runner' into enh/runner

# Conflicts:
#	runner/actions/stream.go
#	runner/runner.go

DawinYurtseven committed:
Import cycle broken for server communications

DawinYurtseven committed:
Deleted unnecessary files

DawinYurtseven committed:
fixed reference errors and Sentry.TranscriptName error

DawinYurtseven committed:
functional streams are now possible with runner. Added two new models for the preventions of crashes and fail saves for runners( Jobs.go and Action.go. for now, the notify runner method only looks for streams like the notifyWorker func but I'm working on the overlooking of jobs on tumlive side with the models and dao files I added

DawinYurtseven committed:
little progress

YiranDuan721 committed:
Add action silence detect (#1338)

* Realised silence detect action

* Apply silence detection in the job

* align method of setting a new value in the context with existing actions

* send silence results to server on success

YiranDuan721 committed:
Enh/runner audio normalization (#1329)

* Add commands for audio normalization; pass 1 realised in a new action

* Pass 1 realised in a new action

* Make sure transcoding with audio normalization does the same to video.

* Integrate audio normalization into transcode. Apply it only when exactly 1 video exists for the stream.

* Pass 1: properly return error message

* Pass 1: 1. synchronise reading from stdin before extracting info; 2. make sure to close everything in case of error

* Audio normalization: log when applied

* Remove the unused action AudioNormalization

---------

Co-authored-by: Dawin <73998830+DawinYurtseven@users.noreply.github.com>

DawinYurtseven committed:
temporary changes with runner Job creation not done yet

DawinYurtseven committed:
temporary changes with runner Job creation not done yet. created the web page but not fully finished the pop up window. working on the models and their schemas.

DawinYurtseven committed:
temporarily push to see full progress

DawinYurtseven committed:
progress of making the Job/Action system of the TUM-Live side.

DawinYurtseven committed:
Merge remote-tracking branch 'origin/dev' into enh/runner

# Conflicts:
#	.golangci.yml
#	cmd/tumlive/tumlive.go
#	go.work.sum
#	tools/stream-signing.go
#	web/admin.go
#	web/template/partial/footer.gohtml
#	web/template/user-settings.gohtml

DawinYurtseven committed:
merge and update from dev branch

DawinYurtseven committed:
stream running without IngestServer

DawinYurtseven committed:
lint fix

DawinYurtseven committed:
also mock

DawinYurtseven committed:
Aktualisieren von stream.go

jetzt sollte tumlive auf die richtige vm zugreifen müssen

DawinYurtseven committed:
adjusted the hls access from edge server and made the proxy work. adjusted the cmd.yaml for less clutter during hls stream

DawinYurtseven committed:
Merge remote-tracking branch 'origin/enh/runner' into enh/runner

# Conflicts:
#	runner/actions/stream.go

DawinYurtseven committed:
did some work for the admin runner page

SebiWrn commented on runner/actions/stream.go:
```diff
@@ -0,0 +1,149 @@
+
+				localhls := fmt.Sprintf("%v/%d/%d/%s/%s/playlist.m3u8", hostname, courseID, streamID, version, end.Format("15-04-05"))
+
+				edgeURL, err := url.Parse(fmt.Sprintf("http://%v:8089/%v",
```
Edge Server should be accessed via https if not localhost

DawinYurtseven replied:
Domain name has been added to the config for variability, with the default being "http://localhost".

SebiWrn commented on runner/actions/stream.go:
```diff
@@ -0,0 +1,149 @@
+
+				localhls := fmt.Sprintf("%v/%d/%d/%s/%s/playlist.m3u8", hostname, courseID, streamID, version, end.Format("15-04-05"))
+
+				edgeURL, err := url.Parse(fmt.Sprintf("http://%v:8089/%v",
```
Edge Server isn't hosted on Port 8089 when not in development (port needs to be a config option)

DawinYurtseven replied:
the config option has been added. mind that the port is rn only one port and if needed to have multiple edge servers, we need to think of a selection option to not swarm one server full

SebiWrn reviewed (commented):
The whole generating part of the URL should be done on TUM-Live rather than on the runner

DawinYurtseven committed:
added edge to cfg for the streams

DawinYurtseven committed:
Merge branch 'dev' into enh/runner

DawinYurtseven committed:
some small adjustments

DawinYurtseven committed:
Merge remote-tracking branch 'origin/enh/runner' into enh/runner

DawinYurtseven committed:
working on the upload part for the stream. plan is to use this upload action to push everything after a hls stream is done or if the runner sees not posted hls files on start

DawinYurtseven committed:
lint fix

DawinYurtseven committed:
work on day

DawinYurtseven committed:
renamed file to lowercase

DawinYurtseven committed:
Started commenting and fixed the build problem with upload by introducing context (temporary bypass, won't be used anyway with current version)

DawinYurtseven committed:
Working on the runner proto consistency and that the streams are working properly. Found an error where the new registered runners with same hostname but different port didn't update the port, leading to no communication between Beta Tum-Live and the runner. I fixed that part and now the streams should at least stream.

Missing: There seems to be a cancel flag set in the context somewhere in the runner when the stream action is done, leading to the notifyStreamEnded to not work properly with the given ctx. I can fix it quickly by not using ctx or setting the flag manually to not cancel but I want to make sure where the cancel flag is set for errors
